### ISSUE TITLE
Explain why we shouldn't use arrow function with computedFn

### ISSUE THREAD

**Issue by blacksteed232** [mobxjs/mobx#2062]
Hi, I have a question regarding mobx's [documentation](https://mobx.js.org/refguide/computed-decorator.html) on `computedFn` (might be more appropriate to ask over at `mobx-utils`, but this has to do with the doc so I'll ask here)

The doc notes that "don't use arrow functions as the `this` would be incorrect."

I have looked at the [implementation](https://github.com/mobxjs/mobx-utils/blob/master/src/computedFn.ts) of `computedFn` but I don't see how `this` could cause issue, as whichever type of function is passed in, you call `apply()` on that function and pass in `this`, which should refer to the scope of the class that holds the function, so I don't see an issue here.

For example, the `getCar` function below should work as intended right? Am I missing something here? 

```
class Store () {
  @observable car = new Map();
  getCar = computedFn((license_plate) => {
    return this.car.get(license_plate);
  })
}
```

**Comment by FredyC** [comment]
Have you actually tried that example if it works? When the `computedFn` is called like that, how it could know about the class instance? I think that `this` used in that function will be wrong, but I just glimpsed over the code and I am not familiar with this util. Best if you can prepare reproduction.

**Comment by urugator** [comment]
Arrow functions ignore `this` passed to `call`, `bind`, `apply`, so you would get a wrong `this` in cases like:
```javascript
const store = {
  getCar: computedFn((license_plate) => {
    return this.car.get(license_plate);
  }) 
}
 
Store.prototype.getCar = computedFn((license_plate) => {
  return this.car.get(license_plate);
})
```
Considering the example (which should work fine with arrow fn), I am not sure whether this was the author's concern...

**Comment by blacksteed232** [comment]
@urugator @FredyC Hey guys, thanks so much, I totally miss the basics there. I have tried to use normal function. Something like this:

```
class Store () {
  @observable cars = new Map<string, Car>();
  getCar: (lp: string) => Car = computedFn(function (lp: string): Car {
    return this.cars.get(lp);
  })
}
```
Typescript will complain the following (I got the same error even for the example from the [doc](https://mobx.js.org/refguide/computed-decorator.html):

```
'this' implicitly has type 'any' because it does not have any type annotation
An outer value of 'this' is shadowed by this container.
```

probably has to do with [this](https://stackoverflow.com/questions/56204346/typescript-error-an-outer-value-of-this-is-shadowed-by-this-container). I'm not so fluent with Typescript yet so I bet you guys will know the answer to this right away.

---

On a related note, my _old_ way of doing `computed` is like below:

```
getCar = (lp: string): Car => computed(() => {
  return this.cars.get(lp);
}).get();
```

but then again, as mentioned in this [thread](https://github.com/mobxjs/mobx/issues/1388#issuecomment-377014896), it looks like this approach does not benefit from memoization like `computedFn` does, am I right?

**Comment by urugator** [comment]
> Am I right?

Yes, it will create new computed at every invocation, therefore the cached value is never used.
The `computed(() => this.x + this.y).get();` is mostly useless. It allows you to create "inlined intermediate single-time expression" over multiple observables, preventing the notification of observer unless the whole expression yield different value.
However you can always move it into "standalone" (persistent) computed and benefit from the cache as well. 
     
In this case you don't need computed at all. Map has a constant access time, no need to cache the value by it's key.

Can't comment on TS, not using it.

**Comment by blacksteed232** [comment]
@urugator Thanks for confirming. So it turns out I looked at the code for `flow` and and `computedFn` and I realize I can do the same trick to make Typescript happy. 

Somebody already mentioned [it](https://github.com/mobxjs/mobx/issues/1769) and the root is due to Typescript problem with using `this` inside of function declared in this [particular](https://stackoverflow.com/questions/48007891/ts2683ts-this-implicitly-has-type-any-because-it-does-not-have-a-type-ann) [way](https://stackoverflow.com/questions/56204346/typescript-error-an-outer-value-of-this-is-shadowed-by-this-container) so one can pass a [_fake_ this parameter](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters) to assert the type for `this`.

Anyway, thank you guys for the insightful response. I learned a lot. I'm closing this one now.
